phpunit/html/wp-html-walker-test.php

```diff
@@ -0,0 +1,1012 @@
+		$w          = new WP_HTML_Walker( $html_input );
+		$w->next_tag( 'input' );
+		$w->set_attribute( 'checked', false );
+		$this->assertSame( $html_input, (string) $w );
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_updates_boolean_attributes_when_string_passed() {
+		$w = new WP_HTML_Walker(
+			'<form action="/action_page.php"><input checked type="checkbox" name="vehicle" value="Bike"><label for="vehicle">I have a bike</label></form>'
+		);
+		$w->next_tag( 'input' );
+		$w->set_attribute( 'checked', 'checked' );
+		$this->assertSame(
+			'<form action="/action_page.php"><input checked="checked" type="checkbox" name="vehicle" value="Bike"><label for="vehicle">I have a bike</label></form>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_works_with_wrongly_nested_tags() {
+		$w = new WP_HTML_Walker(
+			'<span>123<p>456</span>789</p>'
+		);
+		$w->next_tag( 'span' );
+		$w->set_attribute( 'class', 'span-class' );
+		$w->next_tag( 'p' );
+		$w->set_attribute( 'class', 'p-class' );
+		$this->assertSame(
+			'<span class="span-class">123<p class="p-class">456</span>789</p>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_updates_attributes_in_malformed_html() {
+		$w = new WP_HTML_Walker( self::HTML_MALFORMED );
+		$w->next_tag( 'span' );
+		$w->set_attribute( 'id', 'first' );
+		$w->next_tag( 'span' );
+		$w->set_attribute( 'id', 'second' );
+		$this->assertSame(
+			'<div><span id="first" class="d-md-none" Notifications</span><span id="second" class="d-none d-md-inline">Back to notifications</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_removes_attributes_in_malformed_html() {
+		$w = new WP_HTML_Walker( self::HTML_MALFORMED );
+		$w->next_tag( 'span' );
+		$w->remove_attribute( 'Notifications<' );
+		$this->assertSame(
+			'<div><span class="d-md-none" /span><span class="d-none d-md-inline">Back to notifications</span></div>',
+			(string) $w
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 */
+	public function test_unclosed_script_tag_should_not_cause_an_infinite_loop() {
+		$w = new WP_HTML_Walker( '<script>' );
+		$w->next_tag();
+		$this->assertSame( 'script', $w->get_tag() );
+		$w->next_tag();
+	}
+
+	/**
+	 * @ticket 56299
+	 * @dataProvider data_script_state
+	 */
+	public function test_ignores_contents_of_a_script_tag( $script_then_div ) {
+		$w = new WP_HTML_Walker( $script_then_div );
+		$w->next_tag();
+		$this->assertSame( 'script', $w->get_tag() );
+		$w->next_tag();
+		$this->assertSame( 'div', $w->get_tag() );
+	}
+
+	/**
+	 * Data provider for test_ignores_contents_of_a_script_tag().
+	 *
+	 * @return array {
+	 *     @type array {
+	 *         @type string $script_then_div The HTML snippet containing script and div tags.
+	 *     }
+	 * }
+	 */
+	public function data_script_state() {
+		$examples = array();
+
+		$examples['Simple script tag'] = array(
```

Comment:
This is a non-standard structure for data providers. Can this be changed to match other tests in the test suite?

This applies to other data providers in the PR.

Reply:
it's [standard PHPUnit](https://phpunit.readthedocs.io/en/9.5/writing-tests-for-phpunit.html#writing-tests-for-phpunit-data-providers-examples-datatest1-php), provides a name on test case failures, and was done for readability to separate the long lines and independent inputs.

is it worth making the code and tests less readable to conform to the other tests that don't have the same needs?

Reply:
I'm not sure what you mean, the link in your comment shows a direct return of an array, which is what I was proposing.

I should have been more specific, what I was suggesting for consistency with the rest of the test suite is:

```php
public function data_script_state() {
	return array(
		'Simple script tag' => array(
			'<script><span class="d-none d-md-inline">Back to notifications</span></script><div></div>',
		),

		'Simple uppercase script tag' => array(
			'<script><span class="d-none d-md-inline">Back to notifications</span></SCRIPT><div></div>',
		),

		'Script with a comment opener inside should end at the next script tag closer (dash dash escaped state)' => array(
			'<script class="d-md-none"><!--</script><div></div>-->',
		),
		//...
	);
}
```

Reply:
We had it formatted this way before but linter was unhappy with missing whitespaces after keys 😅

Reply:
> the link in your comment shows a direct return of an array, which is what I was proposing.

sorry - that document doesn't have an anchor in the section immediately following the direct array result. it's in the sentence following the example.

<img width="753" alt="Screen Shot 2022-09-22 at 5 51 36 PM" src="https://user-images.githubusercontent.com/5431237/191874753-df340638-0156-4c3e-ac0d-113314b35b03.png">

Reply:
> I'm not sure what you mean, the link in your comment shows a direct return of an array, which is what I was proposing.
> 
> I should have been more specific, what I was suggesting for consistency with the rest of the test suite is:
> 
> ```
> public function data_script_state() {
> 	return array(
> 		'Simple script tag' => array(
> 			'<script><span class="d-none d-md-inline">Back to notifications</span></script><div></div>',
> 		),
> 
> 		'Simple uppercase script tag' => array(
> 			'<script><span class="d-none d-md-inline">Back to notifications</span></SCRIPT><div></div>',
> 		),
> 
> 		'Script with a comment opener inside should end at the next script tag closer (dash dash escaped state)' => array(
> 			'<script class="d-md-none"><!--</script><div></div>-->',
> 		),
> 		//...
> 	);
> }
> ```

It started this way, but the WordPress coding standards turned it into something like this:

```
	public function data_script_state() {
		return array(
			'Script with a comment opener inside should end at the next script tag closer (dash dash escaped state)' => array(
				'<script class="d-md-none"><!--</script><div></div>-->',
			),
			
			'Simple script tag'                                                                                      => array(
				'<script><span class="d-none d-md-inline">Back to notifications</span></script><div></div>',
			),
			
			'Simple uppercase script tag'                                                                            => array(
				'<script><span class="d-none d-md-inline">Back to notifications</span></SCRIPT><div></div>',
			),
		);
	}
```

Which is less readable and also requires adjustments whenever these cases are updated.

Reply:
I'll resolve this for now as it takes up a lot of vertical space. Feel free to unresolve if needed @costdev